### PR TITLE
feat(gestionnaire): Add the possibility to update demands

### PR DESCRIPTION
**Pull request by nkokla**

### CODE REVIEW

File: src/components/DemandDetails/DemandDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        if (!demandId) return;
+        demandsService
+          .updateDemand(demandId, demandUpdate)
+          .then(({ data: record }) => setDemand(record as Demand));
+      },
+    [demand, demandsService]
+  );
+
+  const handleSubmit = useCallback(
```
Comment (desoindx) [inline src/components/DemandDetails/DemandDetails.tsx:57]: Why `handleSubmit`for a comment ? 
and why is it different from `updateDemand`?

File: src/components/DemandDetails/DemandDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      const demandId = demand?.id;
+      if (!demandId) return;
+      demandsService
+        .updateDemand(demandId, demandUpdate)
+        .then(({ data: record }) => setDemand(record as Demand));
+    },
+    [demand, demandsService]
+  );
+
+  // TODO: extract the API call to another component
```
Comment (desoindx) [inline src/components/DemandDetails/DemandDetails.tsx:77]: I don't see wy it has to be in another component ? 
but if so please create a trello ticket

File: src/components/DemandDetails/DemandDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    case 'checkbox':
+      return (
+        <input
+          type="checkbox"
+          name={key}
+          defaultChecked={demand?.[key] as boolean}
+          onChange={onEvent}
+        />
+      );
+    case 'textarea':
```
Comment (desoindx) [inline src/components/DemandDetails/DemandDetails.tsx:27]: Why don't you use the DSFR inputs ?

File: src/components/manager/manager.tsx [hunk 4]
```diff
@@ -1,24 +1,19 @@
...
 const columns = [
+  ...demandRowsParams,
   {
-    name: 'Nom',
-    label: 'Nom',
-    render: (demands: Demand) =>
-      `${demands.Nom}${demands.Prénom ? ` ${demands.Prénom}` : ''}`,
+    name: 'view-more',
+    label: 'Voir plus',
+    render: (demand: Demand) => <Link href={`/demandes/${demand.id}`}>X</Link>,
```
Comment (desoindx) [inline src/components/manager/manager.tsx:15]: Weird to have an X here ?

File: src/core/infrastructure/repository/manager.ts [hunk 5]
```diff
@@ -2,6 +2,12 @@ import db from 'src/db';
 import base from 'src/db/airtable';
 import { Demand } from 'src/types/Summary/Demand';
 
+// TODO: need Memoization
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:5]: Memoization ? in the back ?

File: src/core/infrastructure/repository/manager.ts [hunk 6]
```diff
@@ -11,21 +17,61 @@ export const getGestionnaire = (addresse: string): string | null => {
...
     .select({
       filterByFormula: `Gestionnaire="${user.gestionnaire}"`,
       sort: [{ field: 'Date demandes', direction: 'desc' }],
     })
     .all();
 
-  return results.map((result) => result.fields as Demand);
+  return records.map(
+    (record) => ({ id: record.id, ...record.fields } as Demand)
+  );
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:35]: I don't get the difference ?

File: src/core/infrastructure/repository/manager.ts [hunk 7]
```diff
@@ -11,21 +17,61 @@ export const getGestionnaire = (addresse: string): string | null => {
...
+};
+
+export const getDemand = async (
+  email: string,
+  demandId: string
+): Promise<Demand | Record<string, never> | undefined> => {
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:41]: Weird type... how can it be a Record<string, never> ??

File: src/core/infrastructure/repository/manager.ts [hunk 8]
```diff
@@ -11,21 +17,61 @@ export const getGestionnaire = (addresse: string): string | null => {
...
+  const user = await getUser(email);
+
+  if (!user) {
+    return {};
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:45]: return null, you have no demands !

File: src/core/infrastructure/repository/manager.ts [hunk 9]
```diff
@@ -11,21 +17,61 @@ export const getGestionnaire = (addresse: string): string | null => {
...
+  }
+
+  try {
+    const record = await base(tableNameFcuDemands).find(demandId);
+    return { id: record.id, ...record.fields } as Demand;
+  } catch (err) {
+    console.error(err);
+    return { error: err as never };
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:53]: throw an error

File: src/pages/api/demands/[demandId].ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+export default async function demand(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const { demandId } = req.query;
+  const session = await getSession({ req });
+  if (!session?.user?.email) {
+    return res.status(204).json([]);
```
Comment (desoindx) [inline src/pages/api/demands/[demandId].ts:26]: empty array ?

File: src/pages/api/demands/[demandId].ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  switch (req.method) {
+    case 'GET': {
+      try {
+        const demand = await getDemand(session.user.email, demandId as string);
+        const error = (demand as any)?.error;
+        if (error) {
+          dealErrors(res, error);
+        }
```
Comment (desoindx) [inline src/pages/api/demands/[demandId].ts:36]: will be much simpler if you throw an error before

File: src/pages/api/demands/[demandId].ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      } catch (error) {
+        console.error(error);
+        res.statusCode = 500;
+        return res.json({
+          message: 'internal server error',
+          code: 'Internal Server Error',
+        });
+      }
+    }
+    case 'POST': {
```
Comment (desoindx) [inline src/pages/api/demands/[demandId].ts:47]: should be a PUT

File: src/pages/api/demands/[demandId].ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+      try {
+        const body = req.body;
+        const demand = await updateDemand(
+          session.user.email,
+          demandId as string,
+          body
```
Comment (desoindx) [inline src/pages/api/demands/[demandId].ts:53]: you have to check what's inside the body, with something like joi

File: src/pages/demandes/[demandId].tsx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+export default function Gestionnaire(): JSX.Element {
+  const router = useRouter();
+  const { demandId }: { demandId?: string } = router.query;
+
+  return (
+    <>
+      <Head>
+        <title>France Chaleur Urbaine - Espace gestionnaire</title>
+      </Head>
+      <MainContainer currentMenu="/demandes">
```
Comment (desoindx) [inline src/pages/demandes/[demandId].tsx:19]:
```suggestion
      <MainContainer currentMenu="/gestionnaire">
```

File: src/pages/demandes/[demandId].tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        {!demandId ? (
```
Comment (desoindx) [inline src/pages/demandes/[demandId].tsx:20]: demandId is always defined here

**Review by desoindx — CHANGES_REQUESTED** [review]

File: src/pages/demandes/[demandId].tsx (hunk 15 above)
Comment (nkokla) [reply src/pages/demandes/[demandId].tsx:20]: Not if you directly send the URL: `requests/` without requestId
This is an idea on how to organize our URL (`requests/` > for all requests and `requests/[demandId]` for a specific request) but, we can also do a redirect.

File: src/components/DemandDetails/DemandDetails.tsx (hunk 1 above)
Comment (desoindx) [reply src/components/DemandDetails/DemandDetails.tsx:67]: ???

File: src/core/infrastructure/repository/manager.ts [hunk 16]
```diff
@@ -10,22 +15,60 @@ export const getGestionnaire = (addresse: string): string | null => {
...
+  const user = await getUser(email);
+
+  if (!user) {
+    return {};
+  }
+
+  try {
+    const record = await base(tableNameFcuDemands).find(demandId);
+    return { id: record.id, ...record.fields } as Demand;
+  } catch (err: any) {
```
Comment (desoindx) [inline src/core/infrastructure/repository/manager.ts:50]: catching an error to throw the same is useless

File: src/pages/api/demands/[demandId].ts (hunk 11 above)
Comment (desoindx) [reply src/pages/api/demands/[demandId].ts:51]: you don't have this error anymore.

**Review by desoindx — APPROVED** [review]

File: src/components/Manager/manager.style.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import styled from 'styled-components';
+
+export const Container = styled.div`
+  max-width: 1700px;
```
[the remark this answers is not in the document]
Comment (desoindx) [reply src/components/Manager/manager.style.ts:4]: it's a bad practice not to have a max width. 1700px is already pretty big

File: src/core/infrastructure/repository/manager.ts [hunk 18]
```diff
@@ -2,30 +2,63 @@ import db from 'src/db';
...
+  }
+  return { id: record.id, ...record.fields } as Demand;
+};
+
+export const updateDemand = async (
+  email: string,
+  demandId: string,
+  update: Partial<Demand>
+): Promise<Demand | null> => {
+  const demand = await getDemand(email, demandId);
```
[the remark this answers is not in the document]
Comment (desoindx) [reply src/core/infrastructure/repository/manager.ts:55]: Yes, to make sure that the user updating the demand has the right to do it

File: src/components/Manager/manager.style.ts [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  font-size: 18px;
+  font-weight: bold;
+`;
+
+export const Distance = styled.div`
+  width: 70px;
+`;
+
+export const TableContainer = styled.div`
+  overflow: scroll;
```
[the remark this answers is not in the document]
Comment (desoindx) [reply src/components/Manager/manager.style.ts:30]: no, we want a scroll horizontally and vertically

**Pull request merged**